code/ai/aicode.cpp

```diff
@@ -11719,6 +11726,27 @@ void ai_process_subobjects(int objnum)
 				}
 			}
 
+			// we need to update client data here.
+			if (MULTIPLAYER_CLIENT) {
+				turret_packet network_turret_data;
+				
+				if (Multi_Turret_Manager.get_latest_packet(objp->net_signature, ship_get_subsys_index(pss), &network_turret_data)) {
+					pss->info_from_server_stamp = network_turret_data.time;
+					pss->turret_enemy_objnum = network_turret_data.target_objnum;
+					pss->turret_enemy_sig = (pss->turret_enemy_objnum > -1) ? Objects[pss->turret_enemy_objnum].signature : 0;
+
+					// if the remote instance had a nullptr, first will be false. second holds the actual angle.
+					if (network_turret_data.angle1.first && pss->submodel_instance_1 != nullptr){
+						pss->submodel_instance_1->cur_angle = network_turret_data.angle1.second;
+					}
+
+					if (network_turret_data.angle2.first && pss->submodel_instance_2 != nullptr){
+						pss->submodel_instance_2->cur_angle = network_turret_data.angle2.second;
+					}
```

Comment:
Will this not set cur_angle each frame?
Intuitively, this should be set once when the packet arrives, and from then the client tracks the transmitted target itself based on the turret_enemy_objnum, no?
Or am I missing something?

Reply:
Ah, good question! It will only set it once because, `Multi_Turret_Manager.get_latest_packet()` discards all packets except those that are not yet elapsed. Those that haven't elapsed are added to the `retained` vector that is made the value in the map for that net signature/subsystem pair.